### PR TITLE
Improve detection of CLI flags unintentionally parsed as arguments

### PR DESCRIPTION
**Pull request by pracucci**
#### What this PR does
Last week we had an issue caused by a mis-specified boolean CLI flag. In details: we called `listblocks -user TENANT-ID -backend gcs -gcs.bucket-name BUCKET-NAME -show-deleted true -max-time MAX-TIME`, but `-max-time MAX-TIME` was ignored because golang **boolean** CLI parsing doesn't support the form `-flag x` but only `-flag` or `-flag=x` ([see doc](https://pkg.go.dev/flag#hdr-Command_line_flag_syntax)). This caused `true -max-time MAX-TIME` to be parsed as arguments instead of flags, but `listblocks` didn't complain because no check is done to ensure no arguments are provided (given it doesn't support args).

In this PR I propose to improve the detection of CLI flags unintentionally parsed as arguments:
- Do not allow to directly call `flag.Parse()`, `flag.NArg()`, `flag.Arg()` and `flag.Args()`
- Parse flags and arguments using the newly introduced utility functions. In particular, the function `util.ParseFlagsWithoutArguments()` should be used whenever there are no expected arguments, so that the program will fail if, for any reason, some arguments are parsed

I manually tested it and it would have prevented the `listblocks` issue (described above) being unnoticed.

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by colega — APPROVED** [review]
LGTM, but I also think this could live in `dskit/flagext`

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM

File: pkg/util/flags.go [hunk 1]
```diff
@@ -49,3 +50,23 @@ func WarnDeprecatedConfig(flagName string, logger log.Logger) {
...
+
+// ParseFlagsAndArguments calls Parse() on the input flag.FlagSet and returns the parsed arguments.
+func ParseFlagsAndArguments(f *flag.FlagSet) ([]string, error) {
+	err := f.Parse(os.Args[1:])
+	return f.Args(), err
+}
+
+// ParseFlags calls Parse() on the input flag.FlagSet and enforces no arguments have been parsed.
+// This utility should be called whenever we only expect CLI flags but no arguments.
+func ParseFlags(f *flag.FlagSet) error {
```
Comment (pstibrany) [inline pkg/util/flags.go:62]: [nit] Shall we better signal the purpose in the name of the function? `ParseFlagsWithoutArguments` or `ParseFlagsOnly`.

File: cmd/metaconvert/main.go [hunk 2]
```diff
@@ -51,7 +52,12 @@ func main() {
 		fmt.Fprintln(flag.CommandLine.Output(), "Flags:")
 		flag.PrintDefaults()
 	}
-	flag.Parse()
+
+	// Parse CLI arguments.
+	if err := util.ParseFlags(flag.CommandLine); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
```
Comment (pstibrany) [inline cmd/metaconvert/main.go:59]: we have `exitWithMessage` here.

File: tools/compaction-planner/main.go [hunk 3]
```diff
@@ -50,7 +51,12 @@ func main() {
 	flag.IntVar(&cfg.shardCount, "shard-count", 4, "Shard count")
 	flag.IntVar(&cfg.splitGroups, "split-groups", 4, "Split groups")
 	flag.StringVar(&cfg.sorting, "sorting", compactor.CompactionOrderOldestFirst, "One of: "+strings.Join(compactor.CompactionOrders, ", ")+".")
-	flag.Parse()
+
+	// Parse CLI arguments.
+	if err := util.ParseFlags(flag.CommandLine); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
```
Comment (pstibrany) [inline tools/compaction-planner/main.go:58]: We use `log.Fatalln` in this file.

```suggestion
		log.Fatalln(err.Error())
```

File: tools/list-deduplicated-blocks/main.go [hunk 4]
```diff
@@ -46,7 +46,12 @@ func main() {
 	logger := gokitlog.NewNopLogger()
 	cfg.bucket.RegisterFlags(flag.CommandLine)
 	flag.StringVar(&cfg.userID, "user", "", "User (tenant)")
-	flag.Parse()
+
+	// Parse CLI arguments.
+	if err := util.ParseFlags(flag.CommandLine); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
```
Comment (pstibrany) [inline tools/list-deduplicated-blocks/main.go:53]:
```suggestion
		log.Fatalln(err.Error())
```

File: tools/listblocks/main.go [hunk 5]
```diff
@@ -63,7 +63,12 @@ func main() {
 	flag.Var(&cfg.maxTime, "max-time", "If set, only blocks with MaxTime <= this value are printed")
 	flag.BoolVar(&cfg.useUlidTimeForMinTimeCheck, "use-ulid-time-for-min-time-check", false, "If true, meta.json files for blocks with ULID time before min-time are not loaded. This may incorrectly skip blocks that have data from the future (minT/maxT higher than ULID).")
 	flag.BoolVar(&cfg.showStats, "show-stats", false, "Show block stats (number of series, chunks, samples)")
-	flag.Parse()
+
+	// Parse CLI flags.
+	if err := util.ParseFlags(flag.CommandLine); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
```
Comment (pstibrany) [inline tools/listblocks/main.go:70]:
```suggestion
		log.Fatalln(err.Error())
```

File: tools/tenant-injector/main.go [hunk 6]
```diff
@@ -26,7 +30,12 @@ func main() {
 	flag.StringVar(&cfg.LocalAddress, "local-address", ":8080", "Local address to listen on (host:port or :port).")
 	flag.StringVar(&cfg.RemoteURL, "remote-address", "", "URL of target to forward requests to to (eg. http://domain.com:80).")
 	flag.StringVar(&cfg.TenantID, "tenant-id", "", "Tenant ID to inject to proxied requests.")
-	flag.Parse()
+
+	// Parse CLI arguments.
+	if err := util.ParseFlags(flag.CommandLine); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
```
Comment (pstibrany) [inline tools/tenant-injector/main.go:37]:
```suggestion
		log.Fatalln(err.Error())
```

File: tools/trafficdump/main.go [hunk 7]
```diff
@@ -39,7 +41,11 @@ func main() {
 	assemblersMaxPagesPerConnection := flag.Int("assembler.max-pages-per-connection", 0, "Upper limit on the number of pages buffered for a single connection. If this limit is reached for a connection, the smallest sequence number will be flushed, along with any contiguous data. If <= 0, this is ignored.")
 	httpServer := flag.String("http-listen", ":18080", "Listen address for HTTP server (useful for profiling of this tool)")
 
-	flag.Parse()
+	// Parse CLI arguments.
+	if err := util.ParseFlags(flag.CommandLine); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
```
Comment (pstibrany) [inline tools/trafficdump/main.go:47]:
```suggestion
		log.Fatalln(err.Error())
```

File: tools/tsdb-index-toc/main.go [hunk 8]
```diff
@@ -12,15 +12,27 @@ import (
...
 func main() {
 	// Clean up all flags registered via init() methods of 3rd-party libraries.
 	flag.CommandLine = flag.NewFlagSet(os.Args[0], flag.ExitOnError)
 
-	flag.Parse()
+	// Parse CLI arguments.
+	args, err := util.ParseFlagsAndArguments(flag.CommandLine)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
```
Comment (pstibrany) [inline tools/tsdb-index-toc/main.go:27]:
```suggestion
		log.Fatalln(err.Error())
```

File: tools/tsdb-index-toc/main.go [hunk 9]
```diff
@@ -12,15 +12,27 @@ import (
...
+	}
+
+	if len(args) != 1 {
+		fmt.Fprintf(os.Stderr, "the command expect 1 argument but %d was provided\n", len(args))
+		os.Exit(1)
```
Comment (pstibrany) [inline tools/tsdb-index-toc/main.go:32]:
```suggestion
		log.Fatalf("the command expect 1 argument but %d was provided", len(args))
```

File: tools/undelete_block_gcs/main.go [hunk 10]
```diff
@@ -15,11 +16,18 @@ import (
...
+	"github.com/grafana/mimir/pkg/util"
 )
 
 func main() {
 	concurrency := flag.Int("concurrency", 8, "number of concurrent goroutines")
-	flag.Parse()
+
+	// Parse CLI arguments.
+	if err := util.ParseFlags(flag.CommandLine); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
```
Comment (pstibrany) [inline tools/undelete_block_gcs/main.go:28]:
```suggestion
		log.Fatalln(err.Error())
```

**Review by pstibrany — APPROVED** [review]

File: pkg/util/flags.go (hunk 1 above)
Comment (pracucci) [reply pkg/util/flags.go:62]: Renamed to `ParseFlagsWithoutArguments`.

**Comment by pracucci** [comment]
> LGTM, but I also think this could live in `dskit/flagext`

See: https://github.com/grafana/dskit/pull/341

**Comment by colega** [comment]
> See: https://github.com/grafana/dskit/pull/341

Thanks, I guess now you'll update this one, right?

**Comment by pracucci** [comment]
> > See: [grafana/dskit#341](https://github.com/grafana/dskit/pull/341)
> 
> Thanks, I guess now you'll update this one, right?

That's right. Done in https://github.com/grafana/mimir/pull/5627/commits/8eb7f30e84f44693627751a012df9bef34185ef7

**Review by colega — APPROVED** [review]

**Pull request merged**